spvenvelope.go

```diff
@@ -0,0 +1,153 @@
+package bc
+
+import (
+	"context"
+	"errors"
+	"time"
+
+	"github.com/libsv/go-bk/envelope"
+	"github.com/libsv/go-bt"
+)
+
+// SPVEnvelope is a struct which contains all information needed
+// for a transaction to be verified.
+type SPVEnvelope struct {
+	TxID          string                  `json:"txid"`
+	RawTX         string                  `json:"rawTx,omitempty"`
+	Proof         *MerkleProof            `json:"proof,omitempty"`
+	MapiResponses []envelope.JSONEnvelope `json:"mapiResponses,omitempty"`
+	Inputs        map[string]*SPVEnvelope `json:"inputs"`
+}
+
+// MapiResponse is a callback from mApi
+type MapiResponse struct {
+	CallbackPayload MerkleProof `json:"callbackPayload"`
+	APIVersion      string      `json:"apiVersion"`
+	Timestamp       time.Time   `json:"timestamp"`
+	MinerID         string      `json:"minerId"`
+	BlockHash       string      `json:"blockHash"`
+	BlockHeight     uint64      `json:"blockHeight"`
+	CallbackTxID    string      `json:"callbackTxId"`
+	CallbackReason  string      `json:"callbackReason"`
+}
+
+// VerifyPayment verifies whether or not the txs supplied via the supplied SPVEnvelope are valid
+func (s *SPVClient) VerifyPayment(ctx context.Context, payment *SPVEnvelope) (bool, error) {
+	proofs := make(map[string]bool)
+	outputValues := make(map[string]uint64)
+
+	tx, err := bt.NewTxFromString(payment.RawTX)
+	if err != nil {
+		return false, err
+	}
+
+	valid, err := s.verifyTxs(ctx, payment, nil, true, proofs)
+	if err != nil {
+		return false, err
+	}
+	if !valid {
+		return valid, nil
+	}
+
+	for _, v := range proofs {
+		if !v {
+			return false, errors.New("payment was not verified")
+		}
+	}
+
+	outputValues[tx.GetTxID()] = tx.GetTotalOutputSatoshis()
+
+	return true, nil
+}
+
+func (s *SPVClient) verifyTxs(ctx context.Context, payment *SPVEnvelope, parentInputs []*bt.Input,
+	isRoot bool, proofs map[string]bool) (bool, error) {
+	tx, err := bt.NewTxFromString(payment.RawTX)
+	if err != nil {
+		return false, err
+	}
+	txID := tx.GetTxID()
+	proofs[txID] = false
+
+	if isRoot && payment.Proof != nil {
+		return false, errors.New("root payment must be unconfirmed")
+	}
+
+	for inputTxID, input := range payment.Inputs {
+		if input.TxID == "" {
+			input.TxID = inputTxID
+		}
+		valid, err := s.verifyTxs(ctx, input, tx.GetInputs(), false, proofs)
+		if err != nil {
+			return false, err
+		}
+		if !valid {
+			return valid, nil
+		}
+	}
+
+	// cannot verify the proof or outputs of the root tx
+	if isRoot {
+		proofs[txID] = true
+		return true, nil
+	}
+
+	// if at the leafs of tree and transaction is unconfirmed, fail
+	if (payment.Inputs == nil || len(payment.Inputs) == 0) && payment.Proof == nil {
+		return false, errors.New("no confirmed transaction provided")
```

Comment:
Can can change these errors to named errors, makes it slightly easier for someone integrating and handling specific error cases. Something like ErrNoConfirmed = errors.New("no confirmed transaction provided") etc